### DISCUSSION TITLE
Refactor autotiling

### DISCUSSION THREAD

**Discussion by scherepanov** [gTile/gTile#397] [Ideas]
@schnz - great work on complete gTile re-write! Here are suggestions for next project for gTile.

We need to refactor autotiling. Prime goal is to make very easy for users to write new autotiling functionality and integrate with gTile.
There is a slew of various tiling extensions around gnome-shell. gTile should be an environment of choice to implement any tiling user can imagine.

I am looking to have plugin-like capabilities for tiling in gTile. Here are suggestions.

Project consist of 3 stages.

Stage 1. Abstract autotiling.

Design decision: 
Instead of writing autotiling base class and deriving from it, we will do standalone tiling function interface. Tiling function will be called with all needed info - list of windows and their coordinates, state etc. Tiling function will respond with list of windows to resize, along with new x/y and size x/y.
Goal is to run tiling function in sandbox using sandbox library (JS sandboxing appears to be not broken yet). Users will be able to write new tiling function code in preferences dialog and dynamically execute it, i.e. do new tiling function or modify existing one right out in working gTile, test it and use it.
If JS sandoxing will not work for any reason, we will need to use standard approach with base class for tiling and derived classes for user custom tiling.
gTile autotiling works only inside one monitor. To move window between monitors, users will need to use gnome built-ins.

Tiling function gets following info on call:
1) Monitor config - left/upper coordinates, size x/y, workarea (this excludes various bars - top bar, left bar etc).
2) List of windows on monitor. That includes id, name, x/y and size x/y.
3) State object. Content of object is kept by gTile and given to tiling function on each call. Tiling function got individual state, and can modify it. It is a replacement for class instance.
After call, tiling function can return (may be modified) state object, and list of windows to be resized/moved, along with new x/y and size x/y.

Good test of tiling functions - we should be easily implement Windows (tm)-like keyboard  tiling with Super+Arrows. Looks like it became a standard functionality for all Gnome distributions. Tiling function itself is rather trivial. gTile will handle all hard parts - interface with Gnome to get windows coordinates, get all keyboard events, get result from tiling function and do window move/resize.

Stage 2. Configurable tiling environment. Tiling function publish actions, properties and default keyboard shortcuts, preferences dialog display them on separate tab per tiling function and allow to edit.
Tiling function will publish list of custom events it can receive, along with default keyboard shortcuts.
Preferences dialog will allow to edit map of events to keyboard shortcuts. By default, keyboard shortcuts works only when gTile window is displayed, but can be set global - works without gTile window.
Tiling function will publish list of other properties that should be visible and editable in preferences dialog.
Preferences dialog will get a separate tab per each tiling function, and will display all above properties and keyboard mapping, allow edit all of them or disable tiling function altogether.
Users should be able to add icon for tiling function to gTile window. We should provide a drop-down menu for all enabled tiling functions.

Stage 3. Sandboxed dynamic code execution for tiling function.

Tiling functions are executed in JS sandbox.
Preferences dialog will have additional area where code for tiling function displayed. Users can edit code there and it is executable right away.

Final state. gTile is supplied with ever growing tiling functionality, mostly written by users. We will have some default enabled tiling functions with default config, end user will be able to change to his taste - disable/enable tiling functions, change their properties/keyboard bindings etc.
End users will be able to add new tiling function on a fly in preferences dialog. Edit code in place or get for github/gTile and paste into preferences. Of course, users will be able to contribute their tiling.

**Comment by schnz** [comment]
Hey @scherepanov. I believe I didn't quite fully understand the concept you described. Maybe you could give an example or two on how a practical use-case would look like. In general, I would support the idea to bring in some i3-like capabilities to gnome/gTile that  automatically arrange windows according to user-preferences  whenever new windows are opened/closed.

At the risk of hijacking your issue, I have two other points I wanted your opinion on:
1. You [may saw](https://github.com/gTile/gTile/issues/336#issuecomment-1793706568) that I enhanced the original auto-tiling mechanism by introducing a small DSL I called `GridSpec`. You can also read about it [here](https://github.com/gTile/gTile/tree/next#gridspec-dsl). You can [take a look at the source code](https://github.com/gTile/gTile/blob/2fa4e282debe7ab7d5600dc63a0b1296ee07ae58/src/util/grid.ts#L11-L26) to see the `GridSpec` definitions that resemble the autotile layout of the original extension. Instead of hardcoding them in the source, I plan to make them available through the preferences dialog, thus already allowing the users a great deal more flexibility w.r.t. autotiling. I would like to keep that feature (side-by-side or in addition) to what you proposed.
2. I am currently finalizing the PR and wanted [to ask your permission](https://github.com/gTile/gTile/issues/336#issuecomment-1805601665) to create two new repositories `gTile/gjs-types` and `gTile/gnome-shell-types` that contain the NPM packages required for TypeScript types. I would also like to add you to the [npm gtile organization](https://www.npmjs.com/org/gtile) so that it isn't exclusively owned by me. I would need your npm username for that.

**Comment by scherepanov** [comment]
My policy is all-inclusive consensus OSS process: if developer is contributing to OSS project, he will be able to do whatever he consider would be best for project. Of course we should reach consensus before, but person who actually want to do something gets more weight in reaching consensus.
In short, if you want to do something:

1) Tell everybody that you are going to do
2) listen to other people arguments
3) If you are actually doing something, your opinion will be very honored.
4) If you will be really insisting, you will get it.

Feel free to add new repos to gTile. I created account in npm scherepanov_npm, please add to npm gtile organization.

New syntax for autotiling looks very nice, and cover all cases in original gTile. Great idea! 
This seems to be an error, should be always 1d, for any number of columns:
    1: new GridSpecParser(`cols(1d)`).parse()!,
    2: new GridSpecParser(`cols(2d,2d)`).parse()!,
    3: new GridSpecParser(`cols(3d,3d,3d)`).parse()!,
    4: new GridSpecParser(`cols(4d,4d,4d,4d)`).parse()!,
    5: new GridSpecParser(`cols(5d,5d,5d,5d,5d)`).parse()!,
    6: new GridSpecParser(`cols(6d,6d,6d,6d,6d,6d)`).parse()!,
    7: new GridSpecParser(`cols(7d,7d,7d,7d,7d,7d,7d)`).parse()!,
    8: new GridSpecParser(`cols(8d,8d,8d,8d,8d,8d,8d,8d)`).parse()!,
    9: new GridSpecParser(`cols(9d,9d,9d,9d,9d,9d,9d,9d,9d)`).parse()!,
    10: new GridSpecParser(`cols(1d,1d,1d,1d,1d,1d,1d,1d,1d,1d)`).parse()!,

That brings up suggestion - extend syntax to allow multiplier like cols(1d x 5), equivalent to cols(5d,5d,5d,5d,5d).

I was planning to add same autotiling for rows, but somehow never got to that.

About suggestions for autotiling refactoring in this ticket.
All autotiling functions works same - they got information on input, including current window with focus, list of windows with their sizes/coordinates etc, Original gTile autotiling funcitons get direct access to gnome API to get all this info, I want to abstract this layer. Autotiling functions can keep state. 
Autitiling functions get linked to keyboard shortcuts
When involved, autotiling functions return list of windows to be moved/resized, along with new coordinates x/y and size x/y. Coordinates can be returned in some tiling spec syntax like original gTile or your new autotiling syntax.
gTile gets list of windows with their new coordinates/sizes, and actually do move/resize.

That will abstract autotiling and isolate from Gnome API. Interface to Gnome API is only inside gTile itself.

My other suggestion is to allow autotilng functions to publish info for preferences dialog - what keybindings they want, what options they can have. Preferences dialog will have separate tab with keybinding and properties per autotiling function.

Your new DSL autotiling syntax can be perfectly well implemented in this approach. Let's arbitrarily call it DSL_tiling. DSL_tiling will publish that it will accept property - string with autotiling syntax, and specify that it can be many pairs of keyboard binding - autotiling syntax string. 
Based on published info from DSL_tiling, preferences dialog will add tab, where user will be able to add pairs keyboard binding - string with autotliling syntax.
When user hit key that maps to DSL tiling, gTile will prepare all info form Gnome API - list of windows, their location/sizes, current focus window, and call DSL_tiling. String with autotiling syntax will be passed as well.
DSL_tiling will parse autotiling syntax and produce output list of windows, along with their new coordinates/sizes (in absolute numbers or in tiling syntax) and return to gTile.
gTile will move and resize windows based on returned info.

DSL_tiling will be isolated from Gnome API. 
DSL_tiling, along with any other autotiling, will have a standard API. We can do some autotiling base class and extend DSL_tiling from it. Or, we can try to do dynamic code execution in  sandbox, allow user to enter and edit autotiling fucntion at runtime.

Yes, we will be able to implement i3-like functionality in gTile, using abstraction of autotiling. Abstraction will make implementaiton easier.,

**Comment by schnz** [comment]
>  This seems to be an error, should be always 1d, for any number of columns:

It is more of an ambiguity. `1d` would be working in the same way, i.e., it doesn't matter if a cell has a width of 3/9 or 1/3. I used `cols(3d,3d,3d)` simply as a little visual cue that signals that this is a row consisting of **3** columns.

> That brings up suggestion - extend syntax to allow multiplier like cols(1d x 5), equivalent to cols(5d,5d,5d,5d,5d).

IMO this makes the DSL more complex without adding a real benefit. My goal was to [keep it simple](https://en.wikipedia.org/wiki/KISS_principle) and concise.

---

Regarding your idea: I believe I understood the concept better now, yet I still can't imagine when I would use it in practice (personally). Yet, you know the gTile user base and their needs way better than I do. Technically I suppose I only spot a single problem: I think keybindings can not be added arbitrarily as they [must map to a settings key](https://gjs-docs.gnome.org/meta13~13/meta.display#method-add_keybinding). But there is most certainly a workaround that can be implemented (an internal mapping function of some sort). Oh, and of course the preferences dialog is still missing in the rewrite and would have to be implemented first as a prerequisite.

**Comment by scherepanov** [comment]
Ok, probably it is better to keep your syntax for autotiling.

Yes I saw people asking for various autotiling in past. Also, you can look on quite a few new Gnome-shell extensions on EGO that are doing some kind of tiling. Question is - why all the flavours of autotiling it is not in gTile yet. And the answer - it is relatively labor consuming to add new autotiling to gTile, almost same level of efforts as writing your own extension. There is no benefits to add new kind of autotiling to gTile comparing to writing new extension - practically same efforts. I thought may be we can make adding new autotiling very easy and trivial in gTile, and people will start using gTile to add new tiling capabilities.

Good example of autotiling would be a Windows(tm) built-in - using Super+Arrows you can make window to take any of 4 quarters, left/right half, bottom/up half, full screen or minimized. Last Ubuntu 23.10 has it in Ubuntu tiling extension. Algo is trivial, most implementation efforts are around interfacing with Gnome API and managing keybindings and other properties in preferences dialog. gTile definitely can make job of implementing any autotiling easy, by taking care of everything. User will need to write algo, do some config in preferences, and that's it. User should not know about gnome-shell API and internals.

Nobody can force you to write something you are not agree with. Let's postpone idea of abstracting autotiling, hopefully we will be able to return to it in future. May be I will be able to better explain concept.

**Comment by schnz** [comment]
> Nobody can force you to write something you are not agree with.

Don't get me wrong, **I** wouldn't want to add a new feature to gTile anyway. My initial thought was that it takes me a day or two to port the existing extension to Gnome 45 🤣. We know how that turned out. So I am certainly committed to finish the rewrite and also the preferences dialog (in a second pull request). I learned a lot about the Gnome API which I am grateful for. But after the PR is through I would really like to invest my time into some other stuff again. But I truly believe that the new extension provides a solid architecture that should be easily extendable with the feature you proposed - just not by me. But I would gladly give feedback or do a code review.

> Good example of autotiling would be a Windows(tm) built-in - using Super+Arrows you can make window to take any of 4 quarters, left/right half, bottom/up half, full screen or minimized

Wouldn't that be possible already with the preset mechanism (except for minimizing)?

> There is no benefits to add new kind of autotiling to gTile comparing to writing new extension - practically same efforts. I thought may be we can make adding new autotiling very easy and trivial in gTile, and people will start using gTile to add new tiling capabilities.

Like I said, I am a total newbie to the Gnome ecosystem. 3 weeks ago I wouldn't know a single thing about Gnome extensions (other than how to install them) and I certainly don't know anything about trends in the community w.r.t. tiling-related extensions. So I trust your assessment that there is a demand for this. Again, me personally: If its all the same, I would probably start a new extension if the effort is the same. Just because I like to have tools do "one thing" and do that one thing good. But gTile is already a collection of multiple related features - so adding this capability to gTile would be fine too, I guess.
